php/installer/Database.class.inc

```diff
@@ -44,6 +44,83 @@ class Database
         return $DB;
     }
 
+    /**
+     * Checks if the current MySQL user has a specified privilege.
+     *
+     * @param string $database  The database to check
+     * @param string $table     The table to check
+     * @param string $privilege A string representing the privilege
+     *
+     * @see https://dev.mysql.com/doc/refman/5.5/en/privileges-provided.html
+     *      Table 6.2 Permissible Privileges for GRANT and REVOKE; "Privilege" column
+     *
+     * @return bool Returns true if current MySQL user has the specified privilege
+     */
+    public function hasPrivilege($database, $table, $privilege)
+    {
+        $stmt    = $this->PDO->query("SHOW GRANTS");
+        $results = $stmt->fetchAll(PDO::FETCH_COLUMN);
+        $stmt->closeCursor();
+        $pattern = "/GRANT .*(ALL|${privilege}).* ON" .
+            " `?([*]|${database})`?[.]`?([*]|${table})`?/";
+
+        // Test all privileges return from SHOW GRANTS agains the pattern
+        return array_reduce(
+            $results,
+            function ($carry, $privilege) use ($pattern) {
+                // Return true if a previous iteration returned true or the
+                // current priviledges matches the pattern.
+                return $carry || (preg_match($pattern, $privilege) === 1);
+            },
+            false
+        );
+    }
+
+    /**
+     * Checks if the database exists
+     *
+     * @param bool $database The database name
+     *
+     * @return bool Returns true if the database exists
+     */
+    public function databaseExists($database)
+    {
+        $stmt = $this->PDO->prepare("SHOW DATABASES LIKE :v_db");
+        $stmt->execute(array('v_db' => $database ));
+        $exists = !empty($stmt->fetchAll());
+
+        // this needs to be !empty, because otherwise if there's
+        // no results returned count($dbexists) == 1 when no
+        // results are returned, because the result isn't an array
+        // Something to do with how the Countable interface is
+        // implemented in PHP..
```

Comment:
This comment is on the wrong line since you moved the empty check above it.

(Unrelatedly, I suspect the "Something to do with how the Countable interface is implemented" comment is referring to is this behaviour:  https://wiki.php.net/rfc/counting_non_countables which was changed in 7.2.

"There are no known use cases for this, and it can easily hide bugs")

Reply:
The comment is just confusing. Removed

Reply:
I think the behaviour of count on objects in PHP was super confusing and the comment was trying to explain the workaround it was using..